habitat-lab/habitat/isaacsim/usd_conversion_scene.py

```diff
@@ -0,0 +1,296 @@
+"""This file converts a <EPISODE>.scene_instance.json to a .usd file, 
+using various .glb scene objects to buld the .usd scene instance..
+"""
+
+from pxr import Usd, UsdGeom, Sdf, Gf
+import json
+from pathlib import Path
+import os
+from ssh_conversion import *
+from multiprocessing import Pool, Barrier
+from typing import List, Dict, Union, Tuple
+
+
+# CONSTANTS
+STAGE_INSTANCE = "stage_instance"
+OBJECT_INSTANCES = "object_instances"
+TEMPLATE_NAME = "template_name"
+TRANSLATION = "translation"
+ROTATION = "rotation"
+SCALE = "non_uniform_scale"
+
+
+def find_filepath(filename, parent_directory) -> str:
+    file_found = False
+    for root, dirs, files in os.walk(parent_directory):
+        if filename in files:
+            output_filepath = str(os.path.join(root, filename))
+            file_found = True
+            break
+
+    if not file_found:
+        print(f"{filename} not found in the parent directoy {parent_directory}")
+        return None
+
+    return output_filepath
+
+
+def get_unique_object_names(scene_data) -> list[str]:
+    """
+    This looks at the scene_instance.json by looking at the "template_name" key.
+    HSSD keeps all these objects as .glb. The scene json could use the same type of
+    object multiple times, but we only need the unique list of objects
+    """
+    object_list = scene_data[OBJECT_INSTANCES]  
+
+    object_name_list = []
+    for object in object_list:
+        object_name_list.append(object[TEMPLATE_NAME])
+
+    return list(set(object_name_list))
+
+
+def make_object_filepath_list(
+    glb_data_folder: str,
+    scene_instance_filepath: str,
+    scene_data,   #NOTE: This should be typing for scene_instance.json Dict[str, Union[str, Dict[str, str] , List[ Dict[str, List[float]] ]]],
+    file_extension: str = ".glb",
+) -> List[str]:
+
+    object_name_list = get_unique_object_names(scene_data)
+
+    # Search through the object glb folder to make a list of object filepaths from scene_data
+    object_filepath_list = []
+    for object_name in object_name_list:
+        object_glb = object_name + file_extension
```

Comment:
This is not the right way to find the collision GLB for an object. You need to find the `.object_config.json` file, open that, and extract the `collision_asset` string, which is a filepath to a GLB. We talked about this earlier, and in fact it is mentioned in your pseudocode below around line 144. Here's a half-finished version of scene-instance conversion that I started writing (untested):
```
def find_file(folder, filename):
    result = None
    for root, dirs, files in os.walk(folder):
        if filename in files:
            assert not result
            result = os.path.join(root, filename)
    return result

def convert_hab_scene(scene_filepath):

    from omni.isaac.lab.sim.converters import MeshConverterCfg, MeshConverter

    scenes_folder = os.path.dirname(scene_filepath)
    objects_folder = scenes_folder + "/../objects"
    assert os.path.exists(objects_folder) and os.path.isdir(objects_folder)

    temp_symlink_filename = None # todo "./data/usd/temp_symlink.
    assert False

    import json
    with open(scene_filepath, 'r') as file:
        scene_json_data = json.load(file)    

    assert "object_instances" in scene_json_data
    for obj_instance_json in scene_json_data["object_instances"]:

        object_config_filename = f"{obj_instance_json['template_name']}.object_config.json"
        object_config_filepath = find_file(objects_folder, object_config_filename)
        assert object_config_filepath

        object_config_json_data = None
        with open(object_config_filepath, 'r') as file:
            object_config_json_data = json.load(file)  
        collision_asset_filename = object_config_json_data["collision_asset"]
        collision_asset_filepath = find_file(objects_folder, collision_asset_filename)

        print(f"Converting {collision_asset_filepath}...")
        converter_config = MeshConverterCfg(
            asset_path=collision_asset_filepath,
            usd_dir="./data/usd/objects/",
            usd_file_name=f"{object_config_filename}.usda",
            force_usd_conversion=False,
            collision_approximation="convexDecomposition",
        )
        converter = MeshConverter(converter_config)
        print(f"Wrote {converter.usd_path}")
```

Reply:
My initial draft was to make sure the individual classes imported from isaac lab worked. But you are right, the comments near line 144 were notes from our earlier meeting. The current draft was merely for expediency. Will update this file according to this draft code with object_config.json.